src/adapter/threads.ts

```diff
@@ -659,25 +659,26 @@ export class Thread implements IVariableStoreDelegate {
       return;
     }
 
-    this._pausedDetails = this._createPausedDetails(event);
+    // We store pausedDetails in a local variable to avoid race conditions while awaiting this._smartStepper.shouldSmartStep
+    const pausedDetails = (this._pausedDetails = this._createPausedDetails(event));
     if (!shouldPause) {
       this.resume();
       return;
     }
 
-    if (await this._smartStepper.shouldSmartStep(this._pausedDetails)) {
```

Comment:
You could double check, after awaiting, that `this._pausedDetails === pausedDetails`, so that we don't smart step or trigger onThreadPaused if a race occurs here. (if they changed, just bail out)

Reply:
Changed